core/modules/system/system.module

```diff
@@ -4742,3 +4748,301 @@ function system_transliterate_machine_name($string, $options = array()) {
 
   return $string;
 }
+
+/**
+ * Implements hook_process_html().
+ */
+function system_preprocess_page(&$variables) {
+  backdrop_emit_cache_tags();
+  backdrop_emit_cache_max_age();
+}
+
+/**
+ * Implements hook_ajax_render_alter().
+ */
+function system_ajax_render_alter(&$commands) {
+  backdrop_emit_cache_tags();
+  backdrop_emit_cache_max_age();
+}
+
+/**
+ * Implements hook_emit_cache_tags().
+ */
+function system_emit_cache_tags($tags) {
+  $page_cache_tags = &backdrop_static(__FUNCTION__, array());
+
+  if (config_get('system.core', 'emit_cache_tags')) {
+    backdrop_add_http_header('X-Backdrop-Cache-Tags', implode(' ', $tags));
+  }
+
+  // Store the emitted cache tags for further use.
+  $page_cache_tags = $tags;
+}
+
+/**
+ * Implements hook_emit_cache_max_age().
+ */
+function system_emit_cache_max_age($max_age) {
+  $page_cache_max_age = &backdrop_static(__FUNCTION__, array());
+
+  // Store the emitted cache max_age for further use.
+  $page_cache_max_age = $max_age;
+}
+
+/**
+ * Implements hook_invalidate_cache_tags().
+ */
+function system_invalidate_cache_tags($tags) {
+  // Not using backdrop_static because the detection is based on the running
+  // code, and that is not going to change during execution.
+  static $use_transaction_callback;
+  $deferred_tags = &backdrop_static('deferred_cache_tags', array());
+
+  $connection = Database::getConnection();
+  if ($connection->inTransaction()) {
+    // We are inside a transaction, we must defer until the end of the transaction.
+    foreach ($tags as $tag) {
+      if (empty($deferred_tags)) {
+        $connection->addRootTransactionEndCallback('system_root_transaction_end');
+      }
+      $deferred_tags[$tag] = TRUE;
+    }
+  }
+  else {
+    // Not in a transction, directly invalidate instead.
+    system_do_invalidate_cache_tags($tags);
+  }
+}
+
+/**
+ * Perform cache tag invalidations. Must do outside of a transaction.
+ *
+ * @throws \Exception
+ */
+function system_do_invalidate_cache_tags($tags) {
```

Comment:
We may need to add a check here that the table exists, otherwise someone doing a D7 upgrade, or a Backdrop update, may get a fatal error. Something like:

```
if (!db_table_exists('invalidated_cache_tags')) {
    return;
  }
```